crates/lib/src/compiler/rpcq.rs

```diff
@@ -1,26 +1,52 @@
+//! Provides an RPCQ client for Quilc
+
 use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
 
+use quil_rs::Program;
 use rmp_serde::Serializer;
 use serde::de::DeserializeOwned;
 use serde::{Deserialize, Serialize};
 use uuid::Uuid;
 use zmq::{Context, Socket, SocketType};
 
+use super::quilc;
+
 pub(crate) const DEFAULT_CLIENT_TIMEOUT: f64 = 30.0;
 
 /// A minimal RPCQ client that does just enough to talk to `quilc`
-pub(crate) struct Client {
-    socket: Socket,
+pub struct Client {
+    pub(crate) endpoint: String,
+    socket: Arc<Mutex<Socket>>,
+}
+
+impl Clone for Client {
+    fn clone(&self) -> Self {
+        Self {
+            endpoint: self.endpoint.clone(),
+            socket: self.socket.clone(),
+        }
+    }
```

Comment:
Is this implementation different from `#[derive(Clone)]`? What exactly is the ineficciency you mention in the PR description? Since `socket` is now wrapped in an `Arc`, we are just creating a new reference to it, not initializing a new one, right? That seems like what we would want, but I might be missing something.

Reply:
Yeah. Previously a ZMQ client was created for every request to Quilc. This way we can have a single client.